### PR TITLE
use static libphobos2.a as default

### PR DESCRIPTION
**Pull request by MartinNowak**
- except for OSX where -Bstatic isn't available

### CODE REVIEW

**Comment by MartinNowak** [comment]
Ping, I want to bundle the shared phobos library for freebsd with the next release.

**Comment by MartinNowak** [comment]
Ping again.

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled off

**Comment by DmitryOlshansky** [comment]
> Auto-merge toggled off

NVM wrong PR

**Comment by WalterBright** [comment]
I have some objections to this, as stated in the PR for Linux.

**Comment by MartinNowak** [comment]
> I have some objections to this, as stated in the PR for Linux.

This is needed to keep the static phobos lib default when shipping a shared phobos lib on FreeBSD.

**Comment by MartinNowak** [comment]
Ping, this is needed to __build__ shared libraries on FBSD.

File: src/link.d [hunk 1]
```diff
@@ -622,14 +622,34 @@ extern (C++) int runLINK()
         {
             char* buf = cast(char*)malloc(3 + slen + 1);
             strcpy(buf, "-l");
-            /* Use "-l:libname.a" if the library name is complete
-             */
-            if (slen > 3 + 2 && memcmp(libname, cast(char*)"lib", 3) == 0 && (memcmp(libname + slen - 2, cast(char*)".a", 2) == 0 || memcmp(libname + slen - 3, cast(char*)".so", 3) == 0))
+
+            if (slen > 3 + 2 && memcmp(libname, cast(char*)"lib", 3) == 0)
```
Comment (andralex) [inline src/link.d]: no need for the cast

File: src/link.d [hunk 2]
```diff
@@ -622,14 +622,34 @@ extern (C++) int runLINK()
...
+            {
+                if (memcmp(libname + slen - 2, cast(char*)".a", 2) == 0)
```
Comment (andralex) [inline src/link.d]: no cast needed here, too

File: src/link.d [hunk 3]
```diff
@@ -622,14 +622,34 @@ extern (C++) int runLINK()
...
+                {
+                    argv.push("-Xlinker");
+                    argv.push("-Bstatic");
+                    strncat(buf, libname + 3, slen - 3 - 2);
+                    argv.push(buf);
+                    argv.push("-Xlinker");
+                    argv.push("-Bdynamic");
+                }
+                else if (memcmp(libname + slen - 3, cast(char*)".so", 3) == 0)
```
Comment (andralex) [inline src/link.d]: no cast

**Comment by andralex** [comment]
worksforme after removing the casts. Can't wait to switch to D, that string manipulation is gnarly.

File: src/link.d (hunk 1 above)
Comment (yebblies) [inline src/link.d]: https://issues.dlang.org/show_bug.cgi?id=11837

**Comment by MartinNowak** [comment]
> worksforme after removing the casts. Can't wait to switch to D, that string manipulation is gnarly.

Updated to use `"abc".ptr` instead.

**Comment by yebblies** [comment]
FWIW the existing code used casts because it worked in both C++ and D.  You could also use `slen > 3 + 2 && libname[0..3] == "lib"`

**Comment by MartinNowak** [comment]
> FWIW the existing code used casts because it worked in both C++ and D. You could also use slen > 3 + 2 && libname[0..3] == "lib"

It works using memcmp, can we please move on with this.

**Comment by MartinNowak** [comment]
Ping, still blocked.

**Comment by MartinNowak** [comment]
Could someone please merge this?

**Comment by andralex** [comment]
sorry for the delay!

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**